docs/abi.rst

```diff
@@ -0,0 +1,560 @@
+* :any:`abi.BaseType.encode()` is used to encode a type's value into an encoded byte string.
+* :any:`abi.BaseType.type_spec()` is used to get an instance of :any:`abi.TypeSpec` that describes that type.
+
+:code:`abi.TypeSpec`
+^^^^^^^^^^^^^^^^^^^^
+
+:any:`abi.TypeSpec` is an abstract base class used to describe ABI types. Every child class of :code:`abi.BaseType` also has a companion :code:`abi.TypeSpec` child class. The :code:`abi.TypeSpec` class has a few methods that return information about the type it represents, but one of the class's most important features is the method :any:`abi.TypeSpec.new_instance()`, which creates and returns a new :any:`abi.BaseType` instance of the ABI type it represents.
+
+Static vs Dynamic Types
+^^^^^^^^^^^^^^^^^^^^^^^
+
+An important property of an ABI type is whether it is static or dynamic.
+
+Static types are defined as types whose encoded length does not depend on the value of that type. This property allows encoding and decoding of static types to be more efficient.
+
+Likewise, dynamic types are defined as types whose encoded length does in fact depend on the value that type has. Due to this dependency, the code that PyTeal generates to encode, decode, and manipulate dynamic types is more complex and generally less efficient than the code needed for static types.
+
+Because of the difference in complexity and efficiency when working with static and dynamic types, **we strongly recommend using static types over dynamic types whenever possible**.
+
+Instantiating Types
+^^^^^^^^^^^^^^^^^^^
+
+There are a few ways to create an instance of an ABI type. Each method produces the same result, but some may be more convenient than others.
+
+With the Constructor
+""""""""""""""""""""""
+
+The most obvious way is to use its constructor, like so:
+
+.. code-block:: python
+
+    myUint8 = abi.Uint8()
+    myUint64 = abi.Uint64()
+    myArrayOf12Uint8s = abi.StaticArray(abi.StaticArrayTypeSpec(abi.Uint8TypeSpec(), 12))
+
+For simple types, using the constructor is straightforward and works as you would expect. However, more complex types like :any:`abi.StaticArray` have type-level arguments, so their constructor must take an :any:`abi.TypeSpec` which fully defines all necessary arguments. These types can be created with a constructor, but it's often not the most convenient way to do so.
+
+With an :code:`abi.TypeSpec` Instance
+""""""""""""""""""""""""""""""""""""""
+
+You may remember that :code:`abi.TypeSpec` has a :any:`new_instance() <abi.TypeSpec.new_instance>` method that can be used to instantiate ABI types. This is another way of instantiating ABI types, if you happen to have an :code:`abi.TypeSpec` instance available. For example:
+
+.. code-block:: python
+
+    myUintType = abi.Uint8TypeSpec()
+    myUint8 = myUintType.new_instance()
+
+    myArrayType = abi.StaticArrayTypeSpec(myUintType, 12)
+    myArrayOf12Uint8s = myArrayType.new_instance()
+
+With :code:`abi.make`
```

Comment:
I agree this syntax is more readable. I don't see an obvious way to solve this, but in reality I was annoyed that you made me read the previous section-and-a-half before telling me the nicest way to instantiate a complex type. 

I wonder if a short "tl;dr" section at the top of this document that literally just shows the right examples would be useful for the lazies of the world. Not saying it's a good idea, just thinking.